TorchSharp.PyBridge/PyTorchUnpickler.cs

```diff
@@ -145,27 +152,29 @@ class OrderedDictObjectConstructor : IObjectConstructor {
         /// </summary>
         class TensorObjectConstructor : IObjectConstructor {
             public object construct(object[] args) {
-                // Arg 0: (byte[] data, ScalarType dtype) // returned from our custom pickler
-                var arg0 = (TensorObject)args[0];
-                // Arg 1: storage_offset
-                int storageOffset = (int)args[1];
-                // Arg 2: tensor_shape
-                var shape = ((object[])args[2]).Select(i => (long)(int)i).ToArray();
-                // Arg 3: stride 
-                var stride = ((object[])args[3]).Select(i => (long)(int)i).ToArray();
-                // Arg 4: requires_grad
-                var requiresGrad = (bool)args[4];
+                // Arg 0: returned from our custom pickler
+                var tensorStream = (TensorStream)args[0];
+
+                var constructor = new TensorConstructorArgs {
+                    archiveIndex = tensorStream.archiveIndex,
+                    data = tensorStream.archiveEntry!.Open(),
```

Comment:
Better to store the archive entry itself, and then in the read function to open the stream and close it. 
This way the read function can be called multiple times, and then we don't have any leftover open stream.

Reply:
The problem with storing the `ZipArchiveEntry` is that it gets disposed when the parent `ZipArchive` (line 41) gets disposed within `UnpickleStateDict` (any subsequent `Open()` would fail).  We can delegate the burden of properly disposing of the `ZipArchive` to the caller code, but I think that's too much complexity (and breaks encapsulation) without achieving much--I didn't see a memory performance improvement when I previously tested it.

In the end, I think it's best that the "Unpickler" encapsulates all the logic around dealing with `ZipArchive`.

To double check, I just verified that the `ZipArchiveEntry`'s get disposed when `ZipArchve` gets disposed and this causes the tensor reads to fail.

What do you think?

Reply:
If the `ZipArchiveEntry` is disposed, then the stream becomes unreadable. e.g., try:

```cs
Stream stream;
using (var archive = new ZipArchive(File.OpenRead(@"path/to/sample.zip"))) {
    stream = archive.Entries[0].Open();
}
var value = new StreamReader(stream).ReadToEnd();
```

Therefore, if I understood correctly - if the stream works, then so should the ZipArchiveEntry. 

What do you think?

Reply:
mmm... I can confirm that ZipArchiveEntry.Open() throws an object disposed exception (if we call it downstream) unless I remove the `using` from line 41.

Looking into it more.

Perhaps the "SubReadStream"s that are created from `ZipArchive._archiveStream` keep it from getting disposed?

Reply:
ha! `leaveOpen`

Reply:
Adding `leaveOpen` to the `ZipArchive` constructor makes this example work:

```
Stream stream;
using (var archive = new ZipArchive(File.OpenRead(@"path/to/sample.zip"), ZipArchiveMode.Read, leaveOpen: true)) {
    stream = archive.Entries[0].Open();
}
var value = new StreamReader(stream).ReadToEnd();
```

From the `ZipArchive` implementation:

```
    private void CloseStreams()
    {
      if (!this._leaveOpen)
      {
        this._archiveStream.Dispose();
        this._backingStream?.Dispose();
        this._archiveReader?.Dispose();
      }
      else
      {
        if (this._backingStream == null)
          return;
        this._archiveStream.Dispose();
      }
    }
```

Not sure if that's intended behavior or not.  The archive stream only gets disposed of when a backing stream had to be created.  The backing stream only gets created when the archive stream is not seekable.

Reply:
Yup! Was just writing in as well. 

In that case, I propose the following: 
- Add a check so that if `skipTensorRead` is true, then `leaveOpen` must be true and the input stream must be seekable. 
- In the `TensorConstructor` object, add an internal flag which checks whether the stream was disposed, and if so then throws a readable error when `ReadTensorFromStream` gets called. 

What do you think?
It limits the use-case of the flag, but at least we know it is safe.

Reply:
I think your first proposal makes sense for sure, but it may not be needed after all.  The code I posted is decompiled, so I didn't interpret it right the first time.  Here's a rewrite.

```
        if (this._backingStream != null) {
           // if backing stream exists, dispose original stream
           this._archiveStream.Dispose();
        }
        // otherwise, don't dispose anything.
```

What it is saying is that when `leaveOpen` is set to true, either the backing stream or the archive stream will stay open.  By definition, both of them will be seekable (because when the archive stream is not seekable, the backing stream will be created).

So I think we're good, right?

---

This little rabbit hole helped fill some uncertainties I had about memory management around the zip file...

Reply:
Ah yes, good catch! 
We still need the check that leaveOpen is true though.

Reply:
Good catch!